File: android/core/src/main/kotlin/com/vivianweidai/science/core/api/ApiClient.kt
package com.vivianweidai.science.core.api

import com.vivianweidai.science.core.model.Activity
import com.vivianweidai.science.core.model.ActivityList
import com.vivianweidai.science.core.model.ResearchScience
import com.vivianweidai.science.core.model.ResearchTechResponse
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json

/**
 * Read-only client for activity listings (olympiads + textbooks) and
 * research tech. Source of truth: YAML in public/content/{olympiads,research}/,
 * built to JSON by a Python script, then fetched from vivianweidai.com.
 * No backend, no auth, no writes.
 */
class ApiClient {
    private val mutex = Mutex()
    private val json = Json { ignoreUnknownKeys = true }
    private var cachedActivities: List<Activity>? = null
    private var cachedSciences: List<ResearchScience>? = null

    suspend fun listActivities(): List<Activity> = mutex.withLock {
        cachedActivities?.let { return it }
        val body = Http.getString(OLYMPIADS_URL)
        val list = json.decodeFromString<ActivityList>(body).items
        cachedActivities = list
        list
    }

    suspend fun listResearchSciences(): List<ResearchScience> = mutex.withLock {
        cachedSciences?.let { return it }
        val body = Http.getString(TECH_URL)
        val sciences = json.decodeFromString<ResearchTechResponse>(body).sciences
        cachedSciences = sciences
        sciences
    }

    suspend fun invalidate() = mutex.withLock {
        cachedActivities = null
        cachedSciences = null
    }

    companion object {
        const val OLYMPIADS_URL = "https://vivianweidai.com/olympiads/olympiads.json"
        const val TECH_URL = "https://vivianweidai.com/research/technology.json"
        val shared = ApiClient()
    }
}

File: apple/shared/Core/API/APIClient.swift
import Foundation

/// Read-only client for activity listings (olympiads + textbooks) and the
/// projects wall.
///
/// Source of truth: YAML files under public/content/{olympiads,research}/.
/// A Python build script generates the corresponding .json files in the
/// same dirs, which we fetch from vivianweidai.com directly. No database,
/// no API layer, no writes.
public actor APIClient {
    public static let shared = APIClient()

    public static let olympiadsURL = URL(
        string: "https://vivianweidai.com/olympiads/olympiads.json"
    )!
    public static let wallURL = URL(
        string: "https://vivianweidai.com/projects/thewall/thewall.json"
    )!

    private let session: URLSession
    private let decoder: JSONDecoder
    private var cachedActivities: [Activity]?
    private var cachedWall: TheWallResponse?

    public init(session: URLSession = .shared) {
        self.session = session
        self.decoder = JSONDecoder()
    }

    /// Return cached activities if a previous fetch (e.g. the
    /// app-launch preload) already populated them; otherwise fetch.
    /// Caching here is what makes the background preload useful — when
    /// the user taps the Olympiads tab, its `.task` returns the
    /// already-loaded list instantly.
    public func listActivities() async throws -> [Activity] {
        if let cachedActivities { return cachedActivities }
        let items = try await get(url: Self.olympiadsURL, as: ActivityList.self).items
        cachedActivities = items
        return items
    }

    public func loadWall() async throws -> TheWallResponse {
        if let cachedWall { return cachedWall }
        let wall = try await get(url: Self.wallURL, as: TheWallResponse.self)
        cachedWall = wall
        return wall
    }

    /// Invalidate caches — wired to pull-to-refresh so users can force
    /// a round trip when they've just pushed new YAML.
    public func invalidate() {
        cachedActivities = nil
        cachedWall = nil
    }

    // MARK: - Private

    private func get<T: Decodable>(url: URL, as: T.Type) async throws -> T {
        var request = URLRequest(url: url)
        request.cachePolicy = .reloadRevalidatingCacheData
        let (data, response) = try await session.data(for: request)
        try Self.check(response)
        return try decoder.decode(T.self, from: data)
    }

    private static func check(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse,
              (200..<300).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
    }
}
